src/synapse/workload.py

```diff
@@ -0,0 +1,467 @@
+
+    Returns:
+        Dict: check object converted to its dict representation.
+    """
+    check = Check(CHECK_NGINX_READY_NAME)
+    check.override = "replace"
+    check.level = "ready"
+    check.tcp = {"port": SYNAPSE_NGINX_PORT}
+    return check.to_dict()
+
+
+def check_mjolnir_ready() -> ops.pebble.CheckDict:
+    """Return the Synapse Mjolnir service check.
+
+    Returns:
+        Dict: check object converted to its dict representation.
+    """
+    check = Check(CHECK_MJOLNIR_READY_NAME)
+    check.override = "replace"
+    check.level = "ready"
+    check.http = {"url": f"http://localhost:{MJOLNIR_HEALTH_PORT}/healthz"}
+    return check.to_dict()
+
+
+def execute_migrate_config(container: ops.Container, charm_state: CharmState) -> None:
+    """Run the Synapse command migrate_config.
+
+    Args:
+        container: Container of the charm.
+        charm_state: Instance of CharmState.
+
+    Raises:
+        CommandMigrateConfigError: something went wrong running migrate_config.
+    """
+    _check_server_name(container=container, charm_state=charm_state)
+    # TODO validate if is possible to use SDK instead of command  # pylint: disable=fixme
+    migrate_config_command = [SYNAPSE_COMMAND_PATH, COMMAND_MIGRATE_CONFIG]
+    migrate_config_result = _exec(
+        container,
+        migrate_config_command,
+        environment=get_environment(charm_state),
+    )
+    if migrate_config_result.exit_code:
+        logger.error(
+            "migrate config failed, stdout: %s, stderr: %s",
+            migrate_config_result.stdout,
+            migrate_config_result.stderr,
+        )
+        raise CommandMigrateConfigError(
+            "Migrate config failed, please review your charm configuration"
+        )
+
+
+def enable_metrics(container: ops.Container) -> None:
+    """Change the Synapse configuration to enable metrics.
+
+    Args:
+        container: Container of the charm.
+
+    Raises:
+        EnableMetricsError: something went wrong enabling metrics.
+    """
+    try:
+        config = container.pull(SYNAPSE_CONFIG_PATH).read()
+        current_yaml = yaml.safe_load(config)
+        metric_listener = {
+            "port": int(PROMETHEUS_TARGET_PORT),
+            "type": "metrics",
+            "bind_addresses": ["::"],
+        }
+        current_yaml["listeners"].extend([metric_listener])
+        current_yaml["enable_metrics"] = True
+        container.push(SYNAPSE_CONFIG_PATH, yaml.safe_dump(current_yaml))
+    except ops.pebble.PathError as exc:
+        raise EnableMetricsError(str(exc)) from exc
+
+
+def _get_mjolnir_config(access_token: str, room_id: str) -> typing.Dict:
+    """Create config as expected by mjolnir.
+
+    Args:
+        access_token: access token to be used by the mjolnir bot.
+        room_id: management room id monitored by the Mjolnir.
+
+    Returns:
+        Mjolnir configuration
+    """
+    with open("templates/mjolnir_production.yaml", encoding="utf-8") as mjolnir_config_file:
```

Comment:
What about using `pathlib.Path.read_text`? Pathlib is a more modern interface compared to `open` and you can get what is needed without a context manager

Reply:
Related: #52